libhif/hif-context.c

```diff
@@ -1623,6 +1623,128 @@ hif_context_run(HifContext *context, GCancellable *cancellable, GError **error)
 }
 
 /**
+ * hif_find_package_version: 
+ * @context: a #HifContext instance.
+ * @name: A package name
+ * @pkglist: A GPtrArray used to return packages with matching name and version.
+ *
+ * Helper function that parses package names to find version and release info and
+ * then looks for packages that match all specified criteria.
+ *
+ * Regular expressions are used, first to look for all four parts (name, version,
+ * release, and arch). If there is a partial match, we look for one less part until
+ * we get match. We then use however many parts we found to specify the #HyQuery
+ * used to find matching packages.
+ *
+ * Returns: %TRUE for success, %FALSE otherwise. If successful, @pkglist will contain
+ * the match.
+ *
+ * Since: 0.7.0
+ */
+static gboolean
+hif_find_package_version (HifContext *context, const gchar *name, GPtrArray **pkglist, GError **error)
+{
+    HifContextPrivate *priv = GET_PRIVATE (context);
+    hy_autoquery HyQuery query = NULL;
+    g_autofree gchar* request_name = NULL;
+    gboolean rval = FALSE;
+    GMatchInfo *match_info;
+
+    gchar *n = NULL;
+    gchar *v = NULL;
+    gchar *r = NULL;
+    gchar *a = NULL;
+
+    GRegex *nvra, *nvr, *nv;
+
+    g_debug("* WORK ON NAME: '%s'", name);
+
+    query = hy_query_create(priv->sack);
+    hy_query_filter_in(query, HY_PKG_ARCH, HY_EQ,
+                       (const gchar **) priv->native_arches);
+    hy_query_filter(query, HY_PKG_REPONAME, HY_NEQ, HY_SYSTEM_REPO_NAME);
+    hy_query_filter(query, HY_PKG_ARCH, HY_NEQ, "src");
+
+    /* I don't like hardcoding the archs, but when I use a more generic pattern, sometimes parts of the revision get put in the arch */
+    nvra = g_regex_new("^(.+)-([0-9.]+)-(.+?)\\.(aarch64|alpha|arm|armhfp|i386|ia64|noarch|ppc|ppc64|ppc64le|s390|s390x|sh3|sh4|sparc|x86_64)", 0, 0, error);
```

Comment:
I thought there was probably something like that out there in the large, unfamiliar code base. :-)

I wanted to be able to work with partial nerva, and assumed if there was a function (hy_split_nevra() ) that it probably didn't support that.

In retrospect I'm not sure there actually is a need to support partial names (ex: name + version but none of the other parts)

My day job and family job have both been pretty busy lately, but I will try to find time to look at hy_split_nevra() and try to refactor this code to use it.

Thanks for the feedback!